sycl/source/handler.cpp

```diff
@@ -2012,10 +2012,11 @@ void handler::setUserFacingNodeType(ext::oneapi::experimental::node_type Type) {
 }
 
 std::optional<std::array<size_t, 3>> handler::getMaxWorkGroups() {
-  auto Dev = detail::getSyclObjImpl(detail::getDeviceFromHandler(*this));
+  device Device = detail::getDeviceFromHandler(*this);
+  const auto &DeviceImpl = detail::getSyclObjImpl(Device);
```

Comment:
The `getDeviceFromHandler` function constructs the `device` object from the `device_impl`
and here we need only `device_impl`. I think we can implement one more function (`const DeviceImplPtr &getDeviceImplFromHandler(...)`) to return the `const shared_ptr<device_impl>&`. And we should review all places where the `getDeviceFromHandler` is called and consider replacing it with the `getDeviceImplFromHandler` where appropriate.

Also I am not sure why the `getDeviceFromHandler` function is declared in two headers: `handler.hpp` and `accessor.hpp`